cinema/views.py

```diff
@@ -34,20 +38,72 @@ class MovieViewSet(viewsets.ModelViewSet):
     queryset = Movie.objects.all()
     serializer_class = MovieSerializer
 
+    def get_queryset(self):
+        queryset = Movie.objects.all()
+        if self.action in ["list", "retrieve"]:
+            queryset = queryset.prefetch_related("genres", "actors")
+        if self.action == "list":
+            queryset = self.filter_queryset(queryset)
+        return queryset.distinct()
+
+    def filter_queryset(self, queryset):
+        genres = self.request.query_params.get("genres")
+        if genres:
+            genre_ids = self.convert_to_int_list(genres)
+            queryset = queryset.filter(genres__id__in=genre_ids)
+
+        actors = self.request.query_params.get("actors")
+        if actors:
+            actor_ids = self.convert_to_int_list(actors)
+            queryset = queryset.filter(actors__id__in=actor_ids)
+
+        title = self.request.query_params.get("title")
+        if title:
+            queryset = queryset.filter(title__icontains=title)
+
+        return queryset
+
+    def convert_to_int_list(self, ids):
+        return [int(str_id) for str_id in ids.split(",")]
+
     def get_serializer_class(self):
         if self.action == "list":
             return MovieListSerializer
-
         if self.action == "retrieve":
             return MovieDetailSerializer
-
         return MovieSerializer
 
 
 class MovieSessionViewSet(viewsets.ModelViewSet):
     queryset = MovieSession.objects.all()
     serializer_class = MovieSessionSerializer
 
+    def get_queryset(self):
+        queryset = MovieSession.objects.all()
+
+        movie_id = self.request.query_params.get("movie")
+        if movie_id:
+            queryset = queryset.filter(movie__id=movie_id)
+
+        date = self.request.query_params.get("date")
+        if date:
+            parsed_date = parse_date(date)
+            if parsed_date:
+                queryset = queryset.filter(show_time__date=parsed_date)
+
+        if self.action == "list":
+            queryset = (
+                queryset
+                .annotate(
+                    total_capacity=F(
+                        "cinema_hall__rows") * F("cinema_hall__seats_in_row"),
+                    tickets_sold=Count("tickets"),
+                    tickets_available=F("total_capacity") - F("tickets_sold")
```

Comment:
tickets available and taken seats must be displayed on different actions and different serializers